lib/mongo/socket.rb

```diff
@@ -169,7 +169,22 @@ def eof?
     private
 
     def read_from_socket(length)
-      @socket.read(length) || String.new
+      # Block on data to read for timeout seconds
+      # using the suggested implementation of http://www.ruby-doc.org/core-2.1.3/Kernel.html#method-i-select
+      # to work with SSL connections and some OSs like Ubuntu, CentOS and Red Hat
+      time_left = timeout.to_f
+      data = ""
+      begin
+        while data.length < length
+          data << @socket.read_nonblock(length - data.length)
+        end
+      rescue IO::WaitReadable
+        raise Timeout::Error.new("Took more than #{timeout} seconds to receive data.") if (time_left -= 0.1) <= 0
+        Kernel::select([@socket], nil, [@socket], 0.1)
+        retry
+      end
+
+      data || String.new
```

Comment:
This looks good but I'm going to test on Jenkins so that each version of Ruby, with and without SSL, and various versions of MongoDB are tested.